### ISSUE TITLE
Please add BiDi support to Vim

### ISSUE THREAD

**Issue by GoogleCodeExporter** [vim/vim#204]
```
For those who love using Vim, but need to work in multiple languages at the 
same time, the lack of bi-directional support for working with right-to-left 
languages is a significant drawback.  I understand that this may be difficult 
to implement, but please do consider it as a significant feature for a future 
release of vim and gvim, especially as Unicode use and support is growing in 
popularity.  Thanks for making Vim.
```

Original issue reported on code.google.com by `casey.be...@gmail.com` on 9 Mar 2014 at 8:39

**Comment by Xcess** [comment]
It's really needed. Nearly ALL of modern editors support this feature. +1

**Comment by XVilka** [comment]
I started gathering the summary gist (like I did with true color one) and linked this issue in it https://gist.github.com/XVilka/a0e49e1c65370ba11c17

**Comment by tonymec** [comment]
Full-bidi would IMHO be extremely hard (and therefore error-prone) to implement in Vim, which never was a WYSIWYG text processor but just a plain text editor.

The recommended way to have Vim display full-bidi text is to let the terminal take care of it (which also means running Vim in a bidi-capable terminal and not as gvim). The mlterm terminal can do full-bidi, and Vim is aware of it, but you might have to get its source and compile it yourself: e.g. AFAICT my (openSUSE Linux) distro does not have it in its software repositories.

See also `:help 'termbidi'`

Best regards,
Tony.

**Comment by Houl** [comment]
For now, as a workaround, I'd `:split` (or `:vsplit`) a window on the same buffer, `:set rightleft` in the new window and `'scrollbind'` in both windows.

**Comment by faridcher** [comment]
[neatvi](https://github.com/aligrudi/neatvi/issues/6) already supports it and it works flawlessly for Persian, Arabic, Hebrew, etc. languages even with a non-BIDI terminal. Can we add neatvi BIDI feature to vim?

The author of neatvi has already released a patch for this issue [here](http://vim.1045645.n5.nabble.com/merge-farsi-handling-into-arabic-c-td1210173i20.html) but it seems to be neglected!

THere is also a similar request for Nvim project [here](https://github.com/neovim/neovim/issues/553).

**Comment by XVilka** [comment]
Please note, recently there was created a working group regarding BiDi support in a console:
- https://terminal-wg.pages.freedesktop.org/bidi/
- https://gitlab.freedesktop.org/terminal-wg/bidi

**Comment by XVilka** [comment]
Note, that with the release of [GNOME 3.34](https://www.gnome.org/news/2019/09/gnome-3-34-released/) the support of BiDi by @egmontkob is available in Gnome Terminal out of the box, which makes testing/implementing it in the other programs, such as Vim way easier.

**Comment by faridcher** [comment]
@XVilka I tested the new console with Vim and shared my experience [here](https://github.com/aligrudi/neatvi/issues/8#issuecomment-537115285). The inability to lay out BIDI text from right to left even with `:set rightleft` option in Vim is the main limiting issue. Perhaps @egmontkob could shed light and help here as well.

**Comment by faridcher** [comment]
I have a problem with Vim in a BiDi supported libvte based terminal like gnome-terminal :

Zero width non-joiner (ZWNJ) shows as the unicode escape (<200c>). A workaround is given below which is not great since it hides the ZWNJ and does not join the persian chars:
https://vi.stackexchange.com/questions/17900/how-to-make-vim-display-zwnj-characters-correctly

The problem is also mentioned below even though it is not relevant to kitty terminal:
https://github.com/kovidgoyal/kitty/issues/1851

**Comment by avidseeker** [comment]
Unicode controlling characters like ZWNJ and [Bidi Control](https://en.wikipedia.org/wiki/Unicode_control_characters#Bidirectional_text_control) chars should be completely hidden not concealed. If user wants to show them, such characters should be included within `listchars` so that they can be toggled with `:list`.

**Comment by chrisbra** [comment]
FWIW, I like the fact that a Vim directly shows when a file contains "funny" characters and this is not hidden by default. This has helped me so often debug strange problems.

**Comment by faridcher** [comment]
@chrisbra This could be controlled by an option. These funny characters have meaning in certain languages; if they are not shown correctly, it makes the text ugly and inaccessible to the reader.

**Comment by chrisbra** [comment]
Of course, I was just commenting to the fact that the previous poster claimed that those should be completely hidden by default and not being concealed. And concealing is already possible as mentioned previously.

**Comment by faridcher** [comment]
Well, I agree with the previous poster. ZWNJ should be treated like any other character in my script (Persian). Character concealment even at level=1, has unwanted side effect. I would like to see the plain text (conceal=0) which is great for editing but also see ZWNJ in my text. libvte based terminals and other text editors render it well.

**Comment by chrisbra** [comment]
I had a quick look, but completely hiding probably doesn't make sense (and we don't do this for other characters). If those characters are completely hidden, what should happen when searching for those characters? What about the `go` motion? What happens with line-wrapping?

This all doesn't really make much sense and opens quite a few bit of fundamental questions that Vim is not able to handle very well, remember all the issues around concealing. The best to do is probably to hide those characters as simple blank chars.

**Comment by faridcher** [comment]
@chrisbra good questions. I report my observations in Gnome Terminal: 

Just like Gnome terminal, for go motion, we may stop at the letter right after the hidden char. If we delete this char **backward**, both the char and the hidden char are deleted; if the caret is at the char right before the hidden char and we delete a char **forward**, the hidden char is also deleted.
The ZWNJ role is to make two separate words one single word.Hence the word wrapping should treat the words at either sides of ZWNJ as a single word: 
کتاب ها is two words
کتاب‌ها is single word
I can confirm this behavior in Gedit implementation of Unicode BiDi. open the attached files and try to resize the windows with word-wrapping enabled:
[gedit-zwnj.txt](https://github.com/user-attachments/files/15791735/gedit-zwnj.txt)

**Comment by avidseeker** [comment]
> probably doesn't make sense (and we don't do this for other characters)
 
> questions that Vim is not able to handle very well

If setting this as default confuses users not familiar with Unicode control characters, then do it as an option like `listchars`

> what should happen when searching for those characters

Newlines are hidden characters and they can be searched via regex. Maybe regex should be improved to include Unicode control characters.

**Comment by chrisbra** [comment]
> Newlines are hidden characters and they can be searched via regex.

Newlines are slightly different (and has nothing to do with hidden). They are used to split the input into separate lines to show.
